File: node-gateway/server/services/halt-switch.js
/**
 * [HALT 2026-08-05] Filesystem kill switch — node-gateway half.
 *
 * The mirror of python-engine/halt_switch.py. Both containers mount the same
 * `trading_data` volume at /data, so one `touch /data/HALT` stops every order
 * path in the system regardless of which container owns it.
 *
 * Node may trip the global sentinel when its own broker boundary discovers an
 * execution-authorization failure (for example a non-allow-listed static IP).
 * It never clears the sentinel.
 *
 * FAIL-CLOSED, PRECISELY
 * ----------------------
 * The file's EXISTENCE is the halt; its JSON is attribution only and can never
 * un-halt. `fs.existsSync` is deliberately NOT used: it swallows every error
 * into `false`, so an EACCES or EIO on /data would read as "not halted" and
 * fail OPEN on exactly the storage faults that should stop us cold. We stat
 * directly and treat any error that is not ENOENT as halted.
 *
 * ENTRIES ONLY
 * ------------
 * Callers gate entries. Exits — protective stops, unwinds, GTT legs — are
 * never gated: refusing an exit during a halt strands a live position with no
 * stop, which is strictly worse than whatever tripped the halt.
 */

const fs = require('fs');
const path = require('path');
const { logger } = require('../middleware/logger');

const HALT_DIR = process.env.HALT_DIR || '/data';
const GLOBAL_SENTINEL = 'HALT';

class TradingHaltedError extends Error {
  constructor(attribution) {
    const scope = (attribution && attribution.scope) || 'global';
    const by = (attribution && attribution.by) || 'unknown';
    const reason = (attribution && attribution.reason) || 'no reason recorded';
    super(`trading halted (${scope}) by ${by}: ${reason}`);
    this.name = 'TradingHaltedError';
    this.attribution = attribution || {};
  }
}

/** Sanitise a channel name so it cannot escape HALT_DIR. */
function safeChannel(channel) {
  const cleaned = String(channel).trim().toLowerCase().replace(/[^a-z0-9_-]/g, '');
  if (!cleaned) throw new Error(`channel name is empty after sanitising: ${channel}`);
  return cleaned;
}

function sentinelPath(channel) {
  if (channel === null || channel === undefined) {
    return path.join(HALT_DIR, GLOBAL_SENTINEL);
  }
  return path.join(HALT_DIR, `${GLOBAL_SENTINEL}.${safeChannel(channel)}`);
}

/**
 * Read one sentinel. Returns null when it is definitively absent, or an
 * attribution object when present or indeterminate.
 */
function readSentinel(p) {
  try {
    fs.statSync(p);
  } catch (err) {
    if (err && err.code === 'ENOENT') return null;
    // Cannot determine the kill switch state. Do not trade.
    logger.error({ event_type: 'halt_sentinel_stat_failed', path: p, reason: err.message },
      'halt sentinel unreadable — failing closed');
    return {
      by: 'unknown',
      reason: `halt sentinel unreadable (${err.code || err.message}); failing closed`,
      tripped_at: null,
      unreadable: true,
    };
  }

  let payload = {};
  try {
    const raw = fs.readFileSync(p, 'utf8');
    if (raw.trim()) {
      const parsed = JSON.parse(raw);
      if (parsed && typeof parsed === 'object' && !Array.isArray(parsed)) payload = parsed;
    }
  } catch (err) {
    logger.warn({ event_type: 'halt_sentinel_payload_unparsed', path: p, reason: err.message },
      'halt sentinel present but unparseable — still halted');
  }

  if (payload.by === undefined) payload.by = 'manual_file';
  if (payload.reason === undefined) payload.reason = 'sentinel present with no recorded reason';
  if (payload.tripped_at === undefined) payload.tripped_at = null;
  return payload;
}

/**
 * Is trading halted? The global sentinel is checked first and always wins.
 * @returns {{halted: boolean, attribution: object|null}}
 */
function haltState(channel = null) {
  const globalPayload = readSentinel(sentinelPath(null));
  if (globalPayload !== null) {
    return { halted: true, attribution: { ...globalPayload, scope: 'global' } };
  }

  if (channel !== null && channel !== undefined) {
    let p;
    try {
      p = sentinelPath(channel);
    } catch (err) {
      // A malformed channel must not resolve to "not halted".
      logger.error({ event_type: 'halt_channel_invalid', channel, reason: err.message });
      return {
        halted: true,
        attribution: {
          by: 'unknown',
          reason: `invalid halt channel ${channel}; failing closed`,
          tripped_at: null,
          scope: String(channel),
        },
      };
    }
    const payload = readSentinel(p);
    if (payload !== null) {
      return { halted: true, attribution: { ...payload, scope: safeChannel(channel) } };
    }
  }

  return { halted: false, attribution: null };
}

/** Throw TradingHaltedError when a sentinel is present. Call before entries. */
function assertNotHalted(channel = null) {
  const { halted, attribution } = haltState(channel);
  if (halted) throw new TradingHaltedError(attribution);
}

/** Atomically trip the global halt. First writer wins; never overwrite cause. */
function tripGlobal({ by, reason }) {
  const p = sentinelPath(null);
  const payload = JSON.stringify({
    scope: 'global',
    by: by || 'node_gateway',
    reason: reason || 'broker safety boundary tripped',
    tripped_at: new Date().toISOString(),
  });
  try {
    fs.mkdirSync(HALT_DIR, { recursive: true });
    const fd = fs.openSync(p, 'wx');
    try {
      fs.writeFileSync(fd, payload, 'utf8');
      fs.fsyncSync(fd);
    } finally {
      fs.closeSync(fd);
    }
    logger.error({ event_type: 'global_halt_tripped', path: p, by, reason });
    return true;
  } catch (err) {
    if (err && err.code === 'EEXIST') return false;
    logger.error({ event_type: 'global_halt_trip_failed', path: p, reason: err.message });
    // An unreadable/unwritable safety path already causes haltState to fail
    // closed on subsequent entries; propagate so the caller cannot hide it.
    throw err;
  }
}

/** One-line human summary for /status and Telegram. */
function describe(channel = null) {
  const { halted, attribution } = haltState(channel);
  if (!halted) return 'ARMED - no halt sentinel present';
  const a = attribution || {};
  return `HALTED (${a.scope || 'global'}) since ${a.tripped_at || 'unknown time'} `
       + `by ${a.by || 'unknown'}: ${a.reason || 'no reason recorded'}`;
}

module.exports = {
  TradingHaltedError,
  assertNotHalted,
  haltState,
  describe,
  tripGlobal,
  sentinelPath,
  HALT_DIR,
};

File: python-engine/halt_switch.py
"""[HALT 2026-08-05] Filesystem kill switch, enforced at the broker boundary.

THE DEFECT THIS EXISTS TO FIX
-----------------------------
`performance.check_circuit_breakers` has computed a correct `halted` boolean
since the system was built, and nothing has ever acted on it. The candour is
already in the tree, at partner_orchestrator.py:533:

    # [HONEST-HALT 2026-07-26] This used to broadcast "Our system halted its
    # own trading" ... No entry path is gated by it, so nothing was halted.

Worse, node-gateway -- the container that actually places the momentum orders --
had no concept of a halt at all. The kill switch was a Telegram message.

WHY A FILE AND NOT A DATABASE FLAG
----------------------------------
The halt has to work in exactly the situations where the rest of the system is
not working. A flag in cache.db is unreadable when SQLite is locked by a long
scan; a flag in process memory dies with the process; a flag the LLM has to
respect is not a control at all. A file on the shared volume is checked with one
stat() call, needs no lock, survives a restart, is visible to both containers,
and -- the property that matters at 09:20 on a bad morning -- can be tripped by
hand:

    docker exec python-engine touch /data/HALT

FAIL-CLOSED, PRECISELY
----------------------
The file's EXISTENCE is the halt. Its contents are attribution metadata and
nothing more: a sentinel containing corrupt JSON, an empty sentinel, or one
written by `touch` all halt trading identically. Parsing must never be able to
un-halt.

The subtle half is the error path. `os.path.exists()` returns False when the
stat fails for any reason -- including EACCES and EIO -- so using it here would
fail OPEN on precisely the storage faults that should stop us cold. We stat()
directly and treat every OSError that is not "no such file" as halted: if we
cannot determine the state of the kill switch, we are not trading. That does
mean an unmountable /data halts everything, which is correct; cache.db and the
Kite token live there too, so a system that cannot see /data has no business
sending orders.

SCOPE
-----
A global sentinel (`/data/HALT`) stops every channel. A per-channel sentinel
(`/data/HALT.momentum`) stops one. The global sentinel always wins -- a channel
check consults both, so there is no way to leave a channel live during a global
halt.
"""
from __future__ import annotations

import json
import os
import tempfile
from datetime import datetime, timezone
from typing import Optional

import structlog

logger = structlog.get_logger(__name__)

#: Directory holding the sentinels. Shared docker volume, mounted by both
#: python-engine and node-gateway.
HALT_DIR = "/data"

#: Global sentinel filename. Kept short and uppercase so it is obvious in an
#: `ls /data` during an incident.
GLOBAL_SENTINEL = "HALT"

#: Recognised trip sources, recorded in the payload for the audit trail. Not
#: enforced -- an unknown value is stored as-is rather than rejected, because
#: refusing to record a trip is worse than recording an odd label.
TRIP_SOURCES = ("operator", "circuit_breaker", "watchdog", "manual_file")


class TradingHalted(Exception):
    """Raised at the broker boundary when a sentinel is present.

    Carries the attribution payload so the caller can tell the operator WHY
    trading stopped without re-reading the file.
    """

    def __init__(self, channel: Optional[str], attribution: dict):
        self.channel = channel
        self.attribution = attribution
        reason = attribution.get("reason") or "no reason recorded"
        by = attribution.get("by") or "unknown"
        scope = attribution.get("scope") or "global"
        super().__init__(
            f"trading halted ({scope}) by {by}: {reason}"
        )


def sentinel_path(channel: Optional[str] = None) -> str:
    """Path of the sentinel for `channel`, or the global sentinel when None."""
    if channel is None:
        return os.path.join(HALT_DIR, GLOBAL_SENTINEL)
    safe = _safe_channel(channel)
    return os.path.join(HALT_DIR, f"{GLOBAL_SENTINEL}.{safe}")


def _safe_channel(channel: str) -> str:
    """Reduce a channel name to something that cannot escape HALT_DIR.

    A channel name reaches this module from config and from Telegram commands.
    `/data/HALT.../../etc/passwd` must not be constructible, and a name that
    normalises to empty must not silently become the GLOBAL sentinel -- that
    would turn a typo in a per-channel clear into an unintended global clear.
    """
    cleaned = "".join(c for c in str(channel).strip().lower() if c.isalnum() or c in "_-")
    if not cleaned:
        raise ValueError(f"channel name is empty after sanitising: {channel!r}")
    return cleaned


def _read_sentinel(path: str) -> Optional[dict]:
    """Return the attribution dict when `path` exists, else None.

    Fail-closed: an unreadable-but-present sentinel, or a stat that fails for
    any reason other than absence, yields a dict (i.e. HALTED). Only a clean
    "file is not there" yields None.
    """
    try:
        os.stat(path)
    except FileNotFoundError:
        return None
    except OSError as exc:
        # Cannot determine the kill switch state. Do not trade.
        logger.error("halt_sentinel_stat_failed", path=path, err=str(exc))
        return {
            "by": "unknown",
            "reason": f"halt sentinel unreadable ({exc.__class__.__name__}); failing closed",
            "tripped_at": None,
            "unreadable": True,
        }

    # Present. Contents are attribution only and can never un-halt.
    try:
        with open(path, "r", encoding="utf-8") as fh:
            raw = fh.read(4096)
        payload = json.loads(raw) if raw.strip() else {}
        if not isinstance(payload, dict):
            payload = {}
    except (OSError, ValueError) as exc:
        logger.warning("halt_sentinel_payload_unparsed", path=path, err=str(exc))
        payload = {}

    payload.setdefault("by", "manual_file")
    payload.setdefault("reason", "sentinel present with no recorded reason")
    payload.setdefault("tripped_at", None)
    return payload


def halt_state(channel: Optional[str] = None) -> tuple[bool, Optional[dict]]:
    """Is trading halted for `channel`?

    The global sentinel is consulted first and always wins, so a per-channel
    sentinel can never leave a channel trading during a global halt.

    Returns:
        (halted, attribution). `attribution` is None when not halted, and
        carries a `scope` key of "global" or the channel name when it is.
    """
    payload = _read_sentinel(sentinel_path(None))
    if payload is not None:
        return True, {**payload, "scope": "global"}

    if channel is not None:
        try:
            path = sentinel_path(channel)
        except ValueError as exc:
            # A malformed channel name must not silently resolve to "not
            # halted" -- that is the fail-open we are here to eliminate.
            logger.error("halt_channel_invalid", channel=channel, err=str(exc))
            return True, {
                "by": "unknown",
                "reason": f"invalid halt channel {channel!r}; failing closed",
                "tripped_at": None,
                "scope": str(channel),
            }
        payload = _read_sentinel(path)
        if payload is not None:
            return True, {**payload, "scope": _safe_channel(channel)}

    return False, None


def channel_state(channel: str) -> tuple[bool, Optional[dict]]:
    """State of a channel's OWN sentinel, ignoring the global one.

    `halt_state` deliberately lets the global sentinel win, which is right for
    the enforcement path but wrong for a status display: during a global halt
    every channel would report HALTED and the operator could not see what a
    global clear would leave behind. This answers that question.
    """
    payload = _read_sentinel(sentinel_path(channel))
    if payload is None:
        return False, None
    return True, {**payload, "scope": _safe_channel(channel)}


def assert_not_halted(channel: Optional[str] = None) -> None:
    """Raise `TradingHalted` when a sentinel is present. Call before ordering."""
    halted, attribution = halt_state(channel)
    if halted:
        raise TradingHalted(channel, attribution or {})


def trip(
    reason: str,
    by: str = "operator",
    channel: Optional[str] = None,
) -> dict:
    """Write a sentinel. Idempotent: re-tripping preserves the FIRST trip.

    Preserving the first trip matters because the first reason is the diagnostic
    one. If the daily-loss breaker trips at 10:05 and the drawdown breaker trips
    at 14:20, the operator needs to see 10:05 -- overwriting would erase the
    origin of the incident and leave only its aftershock.

    Raises:
        OSError: if the sentinel cannot be written. A kill switch that silently
            fails to engage is worse than no kill switch, so this is loud.
    """
    path = sentinel_path(channel)

    existing = _read_sentinel(path)
    if existing is not None:
        logger.info(
            "halt_already_tripped",
            path=path, original_by=existing.get("by"),
            original_reason=existing.get("reason"),
        )
        return existing

    payload = {
        "tripped_at": datetime.now(timezone.utc).isoformat(),
        "by": str(by),
        "reason": str(reason)[:500],
        "scope": "global" if channel is None else _safe_channel(channel),
    }
    _atomic_write(path, payload)
    logger.error("halt_tripped", path=path, by=by, reason=reason)
    return payload


def clear(channel: Optional[str] = None) -> bool:
    """Remove a sentinel. Returns True if one was actually removed.

    Clearing the global sentinel does NOT clear per-channel sentinels; each
    scope is cleared explicitly, so "resume everything" cannot be typed by
    accident.
    """
    path = sentinel_path(channel)
    try:
        os.unlink(path)
    except FileNotFoundError:
        return False
    logger.warning("halt_cleared", path=path)
    return True


def _atomic_write(path: str, payload: dict) -> None:
    """Write JSON via temp file + rename.

    Never truncate-then-write. On 2026-07-13 a deploy filled the disk and a
    truncate-on-write zeroed /data/kite_token.json, and the engine ran unarmed
    for a full session. A rename is atomic: the sentinel is either the old
    content or the new one, never an empty file -- and an empty file here would
    still halt, but with no reason recorded.
    """
    directory = os.path.dirname(path) or "."
    fd, tmp = tempfile.mkstemp(dir=directory, prefix=".halt-", suffix=".tmp")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, indent=2, sort_keys=True)
            fh.flush()
            os.fsync(fh.fileno())
        os.replace(tmp, path)
    except BaseException:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


def describe(channel: Optional[str] = None) -> str:
    """One-line human summary for Telegram / status routes."""
    halted, attribution = halt_state(channel)
    if not halted:
        return "ARMED - no halt sentinel present"
    a = attribution or {}
    when = a.get("tripped_at") or "unknown time"
    return (
        f"HALTED ({a.get('scope', 'global')}) since {when} "
        f"by {a.get('by', 'unknown')}: {a.get('reason', 'no reason recorded')}"
    )
